### PR TITLE
Revert "feat(treesitter): add foldtext with treesitter highlighting (#25391)"

### PR DESCRIPTION
**Pull request by tbung**
This reverts commit 9ce1623837a817c3f4f5deff9c8ba862578b6009.

It started with me, it should end with me ;)

Remove `vim.treesitter.foldtext` in favor of https://github.com/neovim/neovim/pull/20750.

### CODE REVIEW

File: runtime/lua/vim/_meta/api_keysets_extra.lua [hunk 1]
```diff
@@ -0,0 +1,167 @@
+--- @meta _
```
Comment (clason) [inline runtime/lua/vim/_meta/api_keysets_extra.lua:1]: That file is a bit weird... 

I think it might be better to make this a regular commit?

**Comment by wookayin** [comment]
Just my two cents, I'm fine with removing this, but I can see some value of this `foldtext()` implementation: it's *composable*. #20750 is a great feature but I don't think one can (yet) easily customize foldtext by appending/prepending additional information in the foldtext, e.g.

```lua
function highlighted_fun(foo, bar, ...)   -- [folded 10 lines]
```

whereas doing this with `vim.treesitter.foldtext()` is pretty straightforward.

File: runtime/lua/vim/_meta/api_keysets_extra.lua (hunk 1 above)
Comment (tbung) [reply runtime/lua/vim/_meta/api_keysets_extra.lua]: Yes, merging did some unexpected changes to this file, I removed the changes, should be fine now! (Also damn you are fast :smile: )

**Comment by tbung** [comment]
It is composable, but I don't think it should be in core. I think this served well as a proof of concept, for core it should not be specific to treesitter. If you are interested I can upload the code again as a plugin, or you can do it yourself if you want.

File: runtime/doc/news.txt [hunk 2]
```diff
@@ -123,7 +123,8 @@ BREAKING CHANGES IN HEAD                                    *news-breaking-dev*
 The following breaking changes were made during the development cycle to
 unreleased features on Nvim HEAD.
 
-• ...
+• Removed `vim.treesitter.foldtext` as transparent foldtext is now the default
+  https://github.com/neovim/neovim/pull/20750
```
Comment (clason) [inline runtime/doc/news.txt:127]: Nit: it's not default, you need to set `foldtext=''` explicitly.

**Comment by wookayin** [comment]
> It is composable, but I don't think it should be in core. I think this served well as a proof of concept, for core it should not be specific to treesitter. If you are interested I can upload the code again as a plugin, or you can do it yourself if you want.

Yes, I agree with you, we can safely remove it from core. What I meant for composability is `foldexpr = ""` (not this `vim.treesitter.foldtext()`.

File: runtime/doc/news.txt (hunk 2 above)
Comment (tbung) [reply runtime/doc/news.txt]: In my head `foldtext=''` was the default, but you are right, I changed it.

**Review by clason — APPROVED** [review]

File: runtime/doc/news.txt (hunk 2 above)
Comment (gpanders) [reply runtime/doc/news.txt]: It is a good candidate to _become_ a default though

**Review by justinmk — APPROVED** [review]
Thank you for following up on this. Avoiding overlapping features/code is very helpful.

**Comment by magnusriga** [comment]
> Just my two cents: While I'm fine with removing this, I can still see some value of this `foldtext()` implementation: it's _composable_. #20750 is a great feature but I don't think one can (yet) easily customize foldtext by appending/prepending additional information in the foldtext with `foldtext = ""`, e.g. into:
> 
> ```lua
> function highlighted_folded_fun(foo, bar, ...) ... end  -- [folded 10 lines]
> ```
> 
> whereas doing this with `vim.treesitter.foldtext()` is pretty straightforward.
> 
> EDIT: After a bit more thinking, I feel this should be better possible by introducing some special string in the foldtext.

I was looking for the same ability. I.e. keep line with highlighting from treesitter, but add some additional text. Did you figure out a way to do it?

Thank you.

**Comment by magnusriga** [comment]
I ended up writing the below, which accurately recreates the highlighting of the first line of the fold, using treesitter.

Including it here, in case useful to others.

Just set it like this: `vim.opt.foldtext =  "v:lua.require'util.ui'.foldtext()"`.

```lua
-- `~/.config/nvim/lua/util/ui.lua`.

function M.foldtext()
  -- Line number of first line of fold when fold is created,
  -- i.e. when `opt.foldtext` is evaluated.
  local pos = vim.v.foldstart

  -- String of first line of fold.
  local line = vim.api.nvim_buf_get_lines(0, pos - 1, pos, false)[1]

 -- Get language of current buffer.
  local lang = vim.treesitter.language.get_lang(vim.bo.filetype)

  -- Create `LanguageTree`, i.e. parser object, for current buffer filetype.
  local parser = vim.treesitter.get_parser(0, lang)

  if parser == nil then
    return vim.fn.foldtext()
  end

  -- Get `highlights` query for current buffer parser, as table from file,
  -- which gives information on highlights of tree nodes produced by parser.
  local query = vim.treesitter.query.get(parser:lang(), "highlights")

  if query == nil then
    return vim.fn.foldtext()
  end

  -- Partial TSTree for buffer, including root TSNode, and TSNodes of folded line.
  -- PERF: Only parsing needed range, as parsing whole file would be slower.
  local tree = parser:parse({ pos - 1, pos })[1]

  local result = {}
  local line_pos = 0
  local prev_range = { 0, 0 }

  -- Loop through matched "captures", i.e. node-to-capture-group pairs, for each TSNode in given range.
  -- Each TSNode could occur several times in list, i.e. map to several capture groups,
  -- and each capture group could be used by several TSNodes.
  for id, node, _ in query:iter_captures(tree:root(), 0, pos - 1, pos) do
    -- Name of capture group from query, for current capture.
    local name = query.captures[id]

    -- Text of captured node.
    local text = vim.treesitter.get_node_text(node, 0)

    -- Range, i.e. lines in source file, captured TSNode spans, where row is first line of fold.
    local start_row, start_col, end_row, end_col = node:range()        Unused local `end_row`.

    -- Include part of folded line between captured TSNodes, i.e. whitespace,
    -- with arbitrary highlight group, e.g. "Folded", in final `foldtext`.
    if start_col > line_pos then
      table.insert(result, { line:sub(line_pos + 1, start_col), "Folded" })
    end

    -- For control flow analysis, break if TSNode does not have proper range.
    if end_col == nil or start_col == nil then
      break
    end

    -- Move `line_pos` to end column of current node,
    -- thus ensuring next loop iteration includes whitespace between TSNodes.
    line_pos = end_col

    -- Save source code range current TSNode spans, so current TSNode can be ignored if
    -- next capture is for TSNode covering same section of source code.
    local range = { start_col, end_col }

    -- Use language specific highlight, if it exists.
    local highlight = "@" .. name
    local highlight_lang = highlight .. "." .. lang
    if vim.fn.hlexist(highlight_lang) then
      highlight = highlight_lang
    end

    -- Insert TSNode text itself, with highlight group from treesitter.
    if range[1] == prev_range[1] and range[2] == prev_range[2] then
      -- Overwrite previous capture, as it was for same range from source code.
      result[#result] = { text, highlight }
    else
      -- Insert capture for TSNode covering new range of source code.
      table.insert(result, { text, highlight })
      prev_range = range
    end
  end

  return result
end
```

**Pull request merged**